polkadot/telemetry/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Telemtetry utils.
+//! 
+//! `telemetry` macro be used from whereever in the Polkadot codebase
+//! in order to send real-time logging information to the telemetry
+//! server (if there is one). We use the async drain adapter of `slog`
+//! so that the logging thread doesn't get held up at all.
+
+extern crate parking_lot;
+extern crate websocket as ws;
+extern crate slog_async;
+extern crate slog_json;
+#[macro_use(o, kv)]
+extern crate slog;
+extern crate slog_scope;
+
+use std::io;
+use parking_lot::Mutex;
+use slog::Drain;
+pub use slog_scope::with_logger;
+
+/// Configuration for telemetry.
+pub struct TelemetryConfig {
+	/// URL of the telemetry WS server.
+	pub url: String,
+	/// What do do when we connect to the server.
+	pub on_connect: Box<Fn() + Send + 'static>,
+}
+
+/// Initialise telemetry.
+pub fn init_telemetry(config: TelemetryConfig) -> slog_scope::GlobalLoggerGuard {
+	let log = slog::Logger::root(
+		slog_async::Async::new(
+			slog_json::Json::default(
+				TelemetryWriter {
+					buffer: vec![],
+					out: Mutex::new(
+						ws::ClientBuilder::new(&config.url).ok().and_then(|mut x| x.connect(None).ok())
+					),
+					config,
+					first_time: true,	// ensures that on_connect will be called.
+				}
+			).fuse()
+		).build().fuse(), o!()
+	);
+	slog_scope::set_global_logger(log)
+}
+
+/// Exactly equivalent to `slog_scope::info`, provided as a convenience.
+#[macro_export]
+macro_rules! telemetry {
+	( $($t:tt)* ) => { $crate::with_logger(|l| slog_info!(l, $($t)* )) }
+}
+
+struct TelemetryWriter {
+	buffer: Vec<u8>,
+	out: Mutex<Option<ws::sync::Client<Box<ws::stream::sync::NetworkStream + Send>>>>,
+	config: TelemetryConfig,
+	first_time: bool,
+}
+
+impl TelemetryWriter {
+	fn ensure_connected(&mut self) {
+		if self.first_time {
+			(self.config.on_connect)();
+			self.first_time = false;
+		}
+		let mut client = self.out.lock();
+		if client.is_none() {
+			*client = ws::ClientBuilder::new(&self.config.url).ok().and_then(|mut x| x.connect(None).ok());
+			drop(client);
+			(self.config.on_connect)();
+		}
+	}
+}
+
+impl io::Write for TelemetryWriter {
+	fn write(&mut self, msg: &[u8]) -> io::Result<usize> {
+		if msg == b"\n" {
+			let _ = self.flush();
```

Comment:
seems like it will never flush unless it gets exactly the string `"\n"` to write. does something else call `flush`?

Reply:
no, but the only thing that writes to it is the json slog module which always provides a single `"\n"` between records. could change it to `contains("\n")` i guess, at the cost of a little speed loss.